crates/rolldown_binding/src/bundler.rs

```diff
@@ -183,12 +183,26 @@ impl Bundler {
   }
 
   fn handle_errors(&self, errs: Vec<BuildDiagnostic>) -> napi::Error {
-    errs.into_iter().for_each(|err| {
-      eprintln!(
-        "{}",
-        err.into_diagnostic_with(&DiagnosticOptions { cwd: self.cwd.clone() }).to_color_string()
-      );
-    });
+    // TODO: is it possible to return as an aggregated error back to js side?
+    // it seems difficult since `Env` is not even available in async fn.
+    // for now, we only surface single js error.
+    // https://github.com/napi-rs/napi-rs/issues/1981#issuecomment-1978208322
+    // https://github.com/napi-rs/napi-rs/issues/945
+    for err in errs {
```

Comment:
I think we could put the napi error at the end of the errs, so we could print as much `diagnostic` as possible, also napi Error usually means something wrong in js side and usually not recoverable, only print one and let user solve them one by one is ok.

Reply:
I just tried a different approach where we can bring back diagnostics on js side via `BindingOutputs.errors` https://github.com/rolldown/rolldown/pull/2759/commits/92e92a81503ae9c654654abd13b95ce8077fa763. In this way, we can bring back native diagnostics to js api as well. For example, this is how syntax error shows up on cli now:

```sh
$ pnpm -C examples/basic-typescript build

> @example/typescript@ build /home/hiroshi/code/others/rolldown/examples/basic-typescript
> rolldown --config ./rolldown.config.js

file:///home/hiroshi/code/others/rolldown/packages/rolldown/dist/shared/src_index-kw1CzMnA.mjs:141
                const errors2 = errors.map((e) => e instanceof Error ? e : Object.assign(new Error(), e));
                                                                                         ^

Error: [PARSE_ERROR] Error: Expected a semicolon or an implicit semicolon after a statement, but found none
   ╭─[index.ts:1:7]
   │
 1 │ syntax error import { hello } from './hello'
   │       │ 
   │       ╰─ 
   │ 
   │ Help: Try insert a semicolon here
───╯

    at file:///home/hiroshi/code/others/rolldown/packages/rolldown/dist/shared/src_index-kw1CzMnA.mjs:141:76
    at Array.map (<anonymous>)
    at transformToRollupOutput (file:///home/hiroshi/code/others/rolldown/packages/rolldown/dist/shared/src_index-kw1CzMnA.mjs:141:26)
    at RolldownBuild.write (file:///home/hiroshi/code/others/rolldown/packages/rolldown/dist/shared/src_index-kw1CzMnA.mjs:2746:10)
    at async bundleInner (file:///home/hiroshi/code/others/rolldown/packages/rolldown/dist/esm/cli.mjs:393:23) {
  kind: 'PARSE_ERROR'
}
```

Reply:
IMO, giving stack trace for a BuildDiagnostic is misleading.

Reply:
Can you try if there are multiple error scenarios?

Reply:
> giving stack trace for a BuildDiagnostic is misleading.

Right, that doesn't make sense, but we can probably strip it before rethrowing on js.

> Can you try if there are multiple error scenarios?

I was wondering that scenario, but couldn't think of it. Is there any example already? I also plan to check how rollup and esbuild show multiple errors.

Reply:
When there are multiple parse errors, it looks like below. Node prints `AggregatedError.errors`, so probably we don't need `eprintln` on rust side. I also stripped stacktrace for native diagnostics. 

```
$ pnpm -C examples/basic-typescript build

> @example/typescript@ build /home/hiroshi/code/others/rolldown/examples/basic-typescript
> rolldown --config ./rolldown.config.js

file:///home/hiroshi/code/others/rolldown/packages/rolldown/dist/shared/src_index-AcwZUxfw.mjs:145
                throw new AggregateError(causes, "Build failed");
                      ^

AggregateError: Build failed
    at transformToRollupOutput (file:///home/hiroshi/code/others/rolldown/packages/rolldown/dist/shared/src_index-AcwZUxfw.mjs:145:9)
    at RolldownBuild.write (file:///home/hiroshi/code/others/rolldown/packages/rolldown/dist/shared/src_index-AcwZUxfw.mjs:2746:10)
    at async bundleInner (file:///home/hiroshi/code/others/rolldown/packages/rolldown/dist/esm/cli.mjs:393:23) {
  [errors]: [
    [Error: [PARSE_ERROR] Error: Unexpected token
       ╭─[hello.ts:1:33]
       │
     1 │ export function hello(): void { :(
       │                                 ┬  
       │                                 ╰── 
    ───╯
    ] {
      kind: 'PARSE_ERROR'
    },
    [Error: [PARSE_ERROR] Error: Expected a semicolon or an implicit semicolon after a statement, but found none
       ╭─[index.ts:1:32]
       │
     1 │ import { hello } from './hello' :(
       │                                │ 
       │                                ╰─ 
       │ 
       │ Help: Try insert a semicolon here
    ───╯
    ] {
      kind: 'PARSE_ERROR'
    }
  ]
}

Node.js v20.18.0
 ELIFECYCLE  Command failed with exit code 1.
```

Reply:
This is a final version of errors on CLI. I used a getter trick from esbuild to avoid expanding `errors` when printed from `console.error`.

```sh
$ pnpm -C examples/basic-typescript build

> @example/typescript@ build /home/hiroshi/code/others/rolldown/examples/basic-typescript
> rolldown --config ./rolldown.config.js

file:///home/hiroshi/code/others/rolldown/packages/rolldown/dist/shared/src_index-BcFJfRrY.mjs:169
                const wrapper = new Error(summary);
                                ^

Error: Build failed with 2 errors:
[PARSE_ERROR] Error: Invalid Character ` `
   ╭─[index.ts:3:6]
   │
 3 │ foo # bar console.log(hello())
   │      ┬  
   │      ╰── 
───╯

[PARSE_ERROR] Error: Expected a semicolon or an implicit semicolon after a statement, but found none
   ╭─[index.ts:3:4]
   │
 3 │ foo # bar console.log(hello())
   │    │ 
   │    ╰─ 
   │ 
   │ Help: Try insert a semicolon here
───╯


    at handleOutputErrors (file:///home/hiroshi/code/others/rolldown/packages/rolldown/dist/shared/src_index-BcFJfRrY.mjs:169:19)
    at transformToRollupOutput (file:///home/hiroshi/code/others/rolldown/packages/rolldown/dist/shared/src_index-BcFJfRrY.mjs:152:2)
    at RolldownBuild.write (file:///home/hiroshi/code/others/rolldown/packages/rolldown/dist/shared/src_index-BcFJfRrY.mjs:2786:10)
    at async bundleInner (file:///home/hiroshi/code/others/rolldown/packages/rolldown/dist/esm/cli.mjs:393:23) {
  errors: [Getter/Setter]
}

Node.js v20.18.0
```